### PR TITLE
[wgsl] Add FAQ for direction on shader language

### PR DESCRIPTION
**Pull request by dneto0**
Describes the decision for WGSL, its main traits.
Avoids rehash of previous arguments.

### CODE REVIEW

**Comment by jeremyong** [comment]
(from the twitter-sphere, but don't worry not coming with pitchforks) I would have liked to have made the meeting but was unable to unfortunately. That said, I do have some questions if it's alright!

Reading the FAQ as stated, is the implication that I will not be able to ship SPIR-V directly and must introduce a "decompilation" step to WGSL prior to ship in my toolchain?

In the examples shown, I'm not seeing a preprocessor. Will there be one? Can WGSL code be modularized and linked?

Moving the precision to a template parameter of each individual type is a somewhat big decision IMO. Is there an easy way of changing the precision for a particular platform/browser/hardware tuple?

Is there a rationale for the heavily rust-flavored syntax when most existing shading languages aren't rust-flavored?

Can you provide some clarity around struct/type declarations? In particular, alignment rules, and how interoperability with SPIR-V hierarchical types will be achieved?

Cheers

**Comment by dj2** [comment]
> Reading the FAQ as stated, is the implication that I will not be able to ship SPIR-V directly and must introduce a "decompilation" step to WGSL prior to ship in my toolchain?

WebGPU will not ingest SPIR-V directly. If your toolchain is currently terminating at SPIR-V, you will have to compile that into WGSL before sending to WebGPU.


> In the examples shown, I'm not seeing a preprocessor. Will there be one? Can WGSL code be modularized and linked?

There is no preprocess for WGSL. The preprocessor has historically been a large area of incompatibilities and bugs.

Not sure what you mean by modularized and linked, it's text, you can concatenate your text together before sending to WebGPU.

> Moving the precision to a template parameter of each individual type is a somewhat big decision IMO. Is there an easy way of changing the precision for a particular platform/browser/hardware tuple?

So, you'd want to send a different shader to Chrome on Mac then Chrome on Windows?


> Is there a rationale for the heavily rust-flavored syntax when most existing shading languages aren't rust-flavored?

The syntax is easier to parse, which makes it harder to parse wrong.


> Can you provide some clarity around struct/type declarations? In particular, alignment rules, and how interoperability with SPIR-V hierarchical types will be achieved?

That is currently an open discussion issue. As spec'd the offsets must be provided for `SSBO` and `UBO` structs and it directly translates to the SPIR-V. That's .... very unfriendly ... so open question is do we want that or something else, and if something else, what ....

**Comment by dneto0** [comment]
Hey @jeremyong, Thanks for your interest!

I'd like to keep this PR focused on landing the bare-bones FAQ as far as it goes.

Your other questions are great but start bleeding into design considerations.  I suggest filing one or more issues with the [wgsl] in the title (and we'll tag with wgsl label).  Then we can address them in a more compartmentalized and complete way.  (And if you don't file the issues, in due time I can fork them off myself.)

**Comment by zenmumbler** [comment]
Just a question from someone building an experimental WebGPU backend into a web 3D engine: given that it says this was agreed upon at a meeting and that the 4 reviewers on this PR are reps for the big 4: will this be the one to unify all vendors? (i.e. no more W(HL)SL or SPIR-V bytecode input to construct GPUShaderModuleDescriptor? (SPIR-V was answered above now, thanks)
Thanks for everyone's efforts here, loved doing a mockup impl for WebGPU.

**Comment by Kangz** [comment]
@zenmumbler yes, it is *the* WebGPU shading language going forward (though it'll take some time to have browsers with some initial implementations).

**Comment by Jasper-Bekkers** [comment]
> It is a text-based language.

It would be nice to have this be explained in the FAQ a bit more and potentially have it contrasted to what WebAssembly is doing (eg. ship binary representation but have a nice human readable textual representation for writing & debugging).

**Comment by jeremyong** [comment]
> WebGPU will not ingest SPIR-V directly. If your toolchain is currently terminating at SPIR-V, you will have to compile that into WGSL before sending to WebGPU.
> There is no preprocess for WGSL. The preprocessor has historically been a large area of incompatibilities and bugs.
> The syntax is easier to parse, which makes it harder to parse wrong.
>Not sure what you mean by modularized and linked, it's text, you can concatenate your text together before sending to WebGPU.

I should mention that part of my motivation for this line of questions was because we have a great deal of code/infrastructure that needs to do all the above and more. Supporting different platforms/OSs/etc requires custom preprocessing, string pasting, macro expansion, etc. so a new language to now parse and/or manipulate to integrate with existing material/vfx editors and tools is definitely a non-trivial burden.

**Comment by dj2** [comment]
In that case, you can generate conforming SPIR-V which we can then convert to WGSL. The WGSL spec has been written with feedback from the process of converting SPIR-V to WGSL.

**Comment by Jasper-Bekkers** [comment]
> In that case, you can generate conforming SPIR-V which we can then convert to WGSL. The WGSL spec has been written with feedback from the process of converting SPIR-V to WGSL.

One has to consider the reality of the situation; which is that shader compiles already are costing hours (a full day for a rebuild isn't unheard of).

**Comment by jeremyong** [comment]
The other fact is that going from IR which is already in SSA form and presumably would save pipeline compilation time back to a high level language strikes me as a bit odd? I can understand providing a new frontend language if the idea is to improve shader authorship workflow (one of the main goals for SPIR-V was to enable this IIRC), but providing a new frontend language to serve as the main ingestion point (which will then subsequently get compiled again) is definitely awkward IMO.

File: design/ShaderLanguageFAQ.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+of the WGSL and WebGPU specifications.
+
+
+## What about my existing shaders, compilers, and other tooling?
+
+This is also expressed as “the last thing I need is another shader language”.
+(See https://xkcd.com/927/)
+
+For developers invested in the SPIR-V ecosystem for Vulkan and OpenGL,
+the key will be convertibility between WGLSL and SPIR-V.
```
Comment (kvark) [inline design/ShaderLanguageFAQ.md]: `WGLSL` typo

Comment (dneto0) [reply design/ShaderLanguageFAQ.md]: Fixed in next commit.

File: design/ShaderLanguageFAQ.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+## Where can I find WGSL tooling?
+
+The WebGPU Community Group does not deliver tooling for WGSL.
+However, WebGPU implementers are expected to develop tooling
+for handling WGSL, including parsing, validation, and compilation.
+
+At this time, we expect at least three essentially independent
+implementations of WGSL ingestion as part of browser support for WebGPU.
+Independent implementation serves as an important check on the quality
```
Comment (litherum) [inline design/ShaderLanguageFAQ.md]: "serves as a check on the quality"? Doesn't that technically indicate that independent implementations reduce quality? (as in "checks and balances")

File: design/ShaderLanguageFAQ.md (hunk 1 above)
Comment (litherum) [inline design/ShaderLanguageFAQ.md]: "The key?" I don't know what that means.

Rather than being aspirational, please be factual. There will almost certainly eventually be open source compilers from many languages, likely including SPIR-V, to WGSL. The WebGPU standardization group will _not_ produce these compilers.

File: design/ShaderLanguageFAQ.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+This is also expressed as “the last thing I need is another shader language”.
+(See https://xkcd.com/927/)
+
+For developers invested in the SPIR-V ecosystem for Vulkan and OpenGL,
+the key will be convertibility between WGLSL and SPIR-V.
+For shaders that only use WebGPU features, WebGPU Community Group
+members expect to develop tooling to faithfully convert between WGSL
+and “Shader” SPIR-V.
+By “faithful”, we mean that all salient properties of a shader should
```
Comment (litherum) [inline design/ShaderLanguageFAQ.md]: Doesn't every compiler (intend to) faithfully preserve behavior for the subset of the source language it intends to support? This paragraph doesn't seem to mean anything, and I would delete it.

File: design/ShaderLanguageFAQ.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+## What about my existing shaders, compilers, and other tooling?
+
+This is also expressed as “the last thing I need is another shader language”.
+(See https://xkcd.com/927/)
+
+For developers invested in the SPIR-V ecosystem for Vulkan and OpenGL,
+the key will be convertibility between WGLSL and SPIR-V.
+For shaders that only use WebGPU features, WebGPU Community Group
+members expect to develop tooling to faithfully convert between WGSL
```
Comment (litherum) [inline design/ShaderLanguageFAQ.md]: Not in the purview of the standardization group, though. I'm not sure what business this sentence has being in the standardization group's repo.

File: design/ShaderLanguageFAQ.md [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+and “Shader” SPIR-V.
+By “faithful”, we mean that all salient properties of a shader should
+be preserved by a round trip conversion between the two ways of writing
+the shader: e.g. it should retain its semantics and its performance
+characteristics.
+Furthermore, those WebGPU Community Group members plan to add an option
```
Comment (litherum) [inline design/ShaderLanguageFAQ.md]: Again, not in the purview of the standardization group. If one member company wants to do something outside the bounds of the standardization group, they can go right ahead. But it isn't relevant to this repo.

File: design/ShaderLanguageFAQ.md (hunk 1 above)
Comment (litherum) [reply design/ShaderLanguageFAQ.md]: I'm not even sure it's worth calling out SPIR-V explicitly. There will almost certainly be many compilers. They assuredly will intend to be faithful.

File: design/ShaderLanguageFAQ.md (hunk 2 above)
Comment (dj2) [reply design/ShaderLanguageFAQ.md]: I don't think so? It's saying that having multiple implementations means we know the spec can be implemented and isn't just a product of a single compiler. Those implementations check that the spec is actually correct, vs what a given implementation happens to do.

File: design/ShaderLanguageFAQ.md (hunk 4 above)
Comment (dj2) [reply design/ShaderLanguageFAQ.md]: The work may not be in the purview, but the question has come up, so having a response to it is good.

File: design/ShaderLanguageFAQ.md (hunk 3 above)
Comment (dj2) [reply design/ShaderLanguageFAQ.md]: This is attempting to say that a round trip of WGSL -> SPIR-V ->WGSL won't be byte for byte identical.

File: design/ShaderLanguageFAQ.md (hunk 5 above)
Comment (dj2) [reply design/ShaderLanguageFAQ.md]: As above, the question comes up, so it is relevant to the FAQ, and what given members are thinking is useful information.

File: design/ShaderLanguageFAQ.md [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+or “WGSL” for short.
+
+It is under development, but will have the following traits:
+
+*   It is a text-based language.
+*   It is suitable for humans to read and write.
+*   It is strongly typed.
+*   It directly represents block-structured programming, including
+    sequence, selection (if/else), and repetition (looping).
+*   It can be faithfully and simply converted to and from SPIR-V
```
[the remark this answers is not in the document]
Comment (magcius) [reply design/ShaderLanguageFAQ.md]: Ignoring the bikeshed over what "simply converted" means, I don't think this should be a goal of WGSL. I thought the working group agreed that WGSL should be defined in terms of SPIR-V semantics. If a conversion to MSL ends up being a challenge, that is the responsibility of the implementer.

File: design/ShaderLanguageFAQ.md [hunk 7]
```diff
@@ -0,0 +1,157 @@
+# FAQ: WebGPU direction on shader language
+
+## What is the shader language for WebGPU?
+
+WebGPU will use a new shader language “WebGPU Shader Language”,
+or “WGSL” for short.
```
[the remark this answers is not in the document]
Comment (jeremyong) [reply design/ShaderLanguageFAQ.md]: Can you describe or point to a write up that expresses your viewpoint on _what_ specifically it is about the web that makes existing instruction sets/languages unsuitable to build off of.

File: design/ShaderLanguageFAQ.md (hunk 1 above)
Comment (kvark) [reply design/ShaderLanguageFAQ.md]: I read the point here not as "convertibility is possible, and there will be compilers from many languages" but as "simple and clean convertibility with SPIR-V specifically is the goal".

File: design/ShaderLanguageFAQ.md (hunk 5 above)
Comment (kvark) [reply design/ShaderLanguageFAQ.md]: > But it isn't relevant to this repo.

True, but in the end we'll just end up with every big participant to engage in an info war by releasing blog posts, where we'd have even less collaboration and mutual agreement on what's being announced...

@litherum do you disagree with the intention here? You don't want to have the existing compilers target WGSL?
If not, we could rephrase this as something like "WebGPU CG members anticipate ..." or "expect that ... will emerge" without mentioning who exactly is going to work on that.

File: design/ShaderLanguageFAQ.md (hunk 2 above)
Comment (Kangz) [reply design/ShaderLanguageFAQ.md]: That's to contrast with WebGL which is a compiler monoculture now.

File: design/ShaderLanguageFAQ.md [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+
+## What browsers support WGSL?
+
+WGSL is the shader language for WebGPU.
+Any browser implementing WebGPU will be required to accept shaders in WGSL.
+Specifically, the WebGPU conformance test suite will use shaders written
+in WGSL.
+
+Browser support for WGSL is expected to be developed over time, along
```
[the remark this answers is not in the document]
Comment (Kangz) [reply design/ShaderLanguageFAQ.md]: Maybe this can be reworded as:

WGSL being a brand new effort, none of the browsers implement it yet. Support for WGSL is expect to replace support for other shading languages in all WebGPU prototypes before WebGPU is released.

File: design/ShaderLanguageFAQ.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (litherum) [reply design/ShaderLanguageFAQ.md]: It absolutely can be defined in terms of SPIR-V semantics. That doesn't mean that it's okay for it to be extremely difficult to compile to HLSL or MSL.

File: design/ShaderLanguageFAQ.md (hunk 2 above)
Comment (litherum) [reply design/ShaderLanguageFAQ.md]: Right, I agree with the intention, but I believe the specific words in this document do not match the intention.

File: design/ShaderLanguageFAQ.md (hunk 1 above)
Comment (litherum) [reply design/ShaderLanguageFAQ.md]: I don't think "simple and clean convertibility" is relevant to the audience of this document. "Simple and clean" is a means to an end; the end is portability, both in terms of behavior and performance. There are other ways of achieving that end.

File: design/ShaderLanguageFAQ.md (hunk 3 above)
Comment (litherum) [reply design/ShaderLanguageFAQ.md]: The round tripping of WGSL -> MSL -> WGSL also won't be byte for byte identical. Also, the round tripping of WGSL -> HLSL -> WGLSL also won't be byte for byte identical.

File: design/ShaderLanguageFAQ.md (hunk 4 above)
Comment (litherum) [reply design/ShaderLanguageFAQ.md]: I'm not disputing having responses to questions. I'm disputing those responses being hosted by the WebGPU standardization group.

File: design/ShaderLanguageFAQ.md (hunk 5 above)
Comment (litherum) [reply design/ShaderLanguageFAQ.md]: Having compilers target this language would be useful and valuable, regardless of the source language. Specifically mentioning a project by one member company which uses this language as a destination language, is irrelevant, and potentially even misleading.

File: design/ShaderLanguageFAQ.md (hunk 4 above)
Comment (magcius) [reply design/ShaderLanguageFAQ.md]: If the availability of a Forth-to-WGSL was a frequently asked question, I think it would totally deserve a mention in the list of Frequently Asked Questions. As it stands, the toolchain support around WGSL and compatibility with other, existing high-level langauges is probably the most frequently asked question about the project, so I think it makes sense to have an answer for it.

File: design/ShaderLanguageFAQ.md (hunk 1 above)
Comment (magcius) [reply design/ShaderLanguageFAQ.md]: The ease of integration with existing shader technologies is *very* much relevant to the audience of this document. You can even see people chiming in about this topic, as a question, *directly* above this comment thread.

File: design/ShaderLanguageFAQ.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (litherum) [reply design/ShaderLanguageFAQ.md]: I can point to a write-up about non-human-writable languages: https://github.com/gpuweb/gpuweb/issues/43

Other members of the group can point you to write-ups about other languages.

**Comment by KTRosenberg** [comment]
One of the big advantages of SpirV was being able to compile existing GLSL shader code asynchronously. Will there at least be an effort to make shaders compilable asynchronously? That includes any translation layers vendors might make.

**Comment by TheSpydog** [comment]
Not addressed in this FAQ is _why_ the group decided on WGSL. I think it’d be beneficial to add a section about why this language was chosen as opposed to direct SPIRV or other alternatives.

File: design/ShaderLanguageFAQ.md (hunk 2 above)
Comment (magcius) [reply design/ShaderLanguageFAQ.md]: WebGL is not a compiler monoculture, and it's actually a major problem for users. Safari's WebGL GLSL parser has bugs I've worked around on past projects. And Chrome-on-Android does not go through ANGLE's shader translator (best I know?), leading to me needing to work around plenty of driver issues in my WebGL applications.

The graphics community has years of experience with the "multiple reimplementations; no reference tooling" state of affairs with GLSL. One of Vulkan's successes in this arena was developing reference tooling, supported by Khronos themselves.

Comment (dneto0) [reply design/ShaderLanguageFAQ.md]: I think the words say what I intend them to say. 

Independent implementation ensures that multiple people have looked at the spec very closely, i.e. close enough to implement it.  Doing so shines a light on lurking ambiguity or gaps in the spec.  It's not a exhaustive proof, but "an important check".  The other line of defense is the test suite.  Hopefully independent implementation will induce ambiguity-reducing discussion, and when that's resolved the decision is baked into both a spec update and a test update.

That's my intent here, but I did't want to write so many words for it.

**Comment by litherum** [comment]
I made proposed changes to this PR [here](https://github.com/gpuweb/gpuweb/pull/576).

I didn't want my proposed changes to be in @dneto0's repo directly, since members of the CG would probably not think to go there to review potential changes to this PR. I also didn't want to just push directly to @dneto0's branch, because that feels kind of rude somehow.

File: design/ShaderLanguageFAQ.md (hunk 3 above)
Comment (dneto0) [reply design/ShaderLanguageFAQ.md]: I mean something less than byte-for-byte identicality, but more than a typical round trip between shader languages.
I hinted at it with the salience of factors affecting "performance".  Details are TBD. But for example, the achievable parallelism should be preserved through the round trip I care about.  E.g. effective occupancy of a warp on NV hardware or wavefront on AMD hardware.  It's likely that a  good proxy for that is private register count, and the hypothesis is that can be faithfully counted by staring at WGSL source.  The goal is to avoid having a translation to/from WGSL induce falling off a performance cliff.  In my view that's critical to the success of WebGPU.

That's why the paragraph exists.

File: design/ShaderLanguageFAQ.md (hunk 5 above)
Comment (dneto0) [reply design/ShaderLanguageFAQ.md]: I didn't mention any specific projects, or particular member companies. I believe I already provided caveats and qualified language as @kvark suggested.
I mentioned an ecosystem that is highly relevant going forward, and was the proposed technology prior to the decision (which makes the question highly relevant).

Hey @litherum, would you prefer if I prefaced this section with a prominent notice along the lines of: "This section is outside the purview of the community group" ?

**Comment by dneto0** [comment]
> I made proposed changes to this PR [here](https://github.com/gpuweb/gpuweb/pull/576).
> 
> I didn't want my proposed changes to be in @dneto0's repo directly, since members of the CG would probably not think to go there to review potential changes to this PR. I also didn't want to just push directly to @dneto0's branch, because that feels kind of rude somehow.

Ah, I didn't see this in time.  I pushed an update which incorporated much of your feedback, but not all.  Let me take a look. Also, consider taking a look at mine.

**Comment by litherum** [comment]
I will rebase mine on top of yours.

File: design/ShaderLanguageFAQ.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (dneto0) [reply design/ShaderLanguageFAQ.md]: Please see latest commit

File: design/ShaderLanguageFAQ.md [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+## Where can I find the spec?
+
+The WGSL spec will be developed on the WebGPU Community Group area on GitHub.
+See https://github.com/gpuweb/gpuweb/tree/master/wgsl
+
+
+## Where do I provide feedback on WGSL?
+
+To provide feedabck on WGSL, file an issue on the spec repo site:
```
Comment (doug-moen) [inline design/ShaderLanguageFAQ.md]:
```suggestion
To provide feedback on WGSL, file an issue on the spec repo site:
```

Comment (dneto0) [reply design/ShaderLanguageFAQ.md]: fixed in next commit.

**Comment by dneto0** [comment]
>  The concerns of developers using SPIR-V today are valid, and the concerns of developers using MSL and HLSL today are just as valid.

I agree.  This is a case of me not wanting to speak for other constituencies.  I invite you or others to add the colour commentary you think is fitting and informative to your intended audience.   

My last paragraph deliberately scoped: it was predicated with "For developers invested in the SPIR-V ecosystem ...".   Stylistically I disliked jamming all that content into one paragraph, but wanted to keep the clear scoping.

**Comment by pjoe** [comment]
Just curious, why was a textual format chosen, if I'm understanding correctly? Especially for web the whole asm.js > WebAssembly has shown the power of binary (smaller size, less parsing time), also won't each browser vendor now have to implement compilers for each platform: WGSL > (SPIR-V, MSL, HLSL)? I may have misunderstood but wasn't the whole idea of the newer low leve API's (Vulkan, Metal, DX12) to minimize having multiple drivers implement complex things like shader compilers?

Just my 2 cents

**Comment by dj2** [comment]
There are pros and cons for both text and binary. In the end, the community group chose text. The asm vs webassembly isn't really an equal comparison as you'll still have HTML, JS and CSS pages which are all written in text.

**Comment by pjoe** [comment]
Hmm not sure I follow?

If I write my 3D app in say C++ or Rust and compile to WebAssembly, then there will only be a very small amount of HTML/JS that the browser need to download/parse. My main app code will be in WebAssambly. However I may have quite a large amount of shader code - that I can imagine would take the browser more time to download/parse as text.

Not trying to pick a fight, just to understand the reasoning behind this.

**Comment by dj2** [comment]
I don't know if we have any numbers on the time it will take to download and parse WGSL code at this point.

**Comment by pjoe** [comment]
Guess it depends a lot on amount and complexity of shaders :)

I just remember back in the day when working with the linux port of unreal how it took like 30 minutes for it to cross compile shaders from HLSL to GLSL ... would be wonderful to avoid anything like that.

**Comment by dj2** [comment]
Even if the input was binary, we have to compile back to Metal Shading Language or HLSL. Only on the case of running on Vulkan could you avoid the compile, but then you still have to run optimization passes to make sure the SPIR-V is for the WebGPU SPIR-V environment.

**Comment by pjoe** [comment]
Right :)

I guess we'll see minification of WGSL for large shader sources .. together with gzip transfer probably makes download size pretty close to binary

**Comment by doug-moen** [comment]
@pjoe To quote aras-p on github, "SPIR-V is a very verbose format, several times larger than same programs expressed in other shader formats (e.g. ... GLSL). The SSA-form with ever increasing IDs is not very appreciated by regular data compressors either. " [ref: github.com/aras-p/smol-v]

So don't assume that because SPIR-V is binary, that it is more compact than WGSL. The text format of WGSL could turn out to be a net performance win due to smaller file sizes and due to program text being more compressible.

**Comment by dneto0** [comment]
Hi @doug-moen  

The issues you point out have been discussed by the WebGPU community group before.  This PR intentionally stayed away from the _why_ aspect.

(To very briefly answer the SPIR-V size question:  I recommend Aras' SMOL-V to compress your shaders.  It's great.  See more info at https://aras-p.info/blog/2018/10/31/SPIR-V-Compression-SMOL-vs-MARK/ )

File: design/ShaderLanguageFAQ.md (hunk 2 above)
Comment (othermaciej) [reply design/ShaderLanguageFAQ.md]: In typical English idiom, "serves as an important check on X" means placing a limitation on X or reducing X, rather than providing evidence for X. c.f. "Congress serves as an important check on the power of the presidency." Perhaps it can have that other meaning too, but in that case why use an ambiguous phrase? You could say "Independent implementation provides evidence for the quality of the spec."

This may seem nitpicky but use of ambiguous wording like this can be legitimately confusing to some people (e.g. non-native English speakers or the neurodiverse).

**Comment by litherum** [comment]
https://webkit.org/blog/9528/webgpu-and-wsl-in-safari/ demonstrates SPIR-V being both slower to compiler and having larger gzipped wire sizes than WSL (a textual shading language).

**Comment by grorg** [comment]
Discussed at [2020-03-10 Teleconference](https://docs.google.com/document/d/10yBasmrNdgvvd44KWC5a3ZQGzjJCjUBoqzZdByC_u7Y/)

**Comment by dneto0** [comment]
Closing in favour of #687.
